src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -1049,14 +1059,20 @@ public void gossipSnitchInfo()
         Gossiper.instance.addLocalApplicationState(ApplicationState.RACK, StorageService.instance.valueFactory.rack(rack));
     }
 
-    public synchronized void joinRing() throws IOException
-    {
+    public synchronized void joinRing(Collection<String> initalTokens) throws IOException {
         if (!joined)
         {
             logger.info("Joining ring by operator request");
             try
             {
-                joinTokenRing(0);
+                if(initalTokens.isEmpty()) {
+                    joinTokenRing(0);
+                } else {
+                    initalTokens.stream().forEach(getPartitioner().getTokenFactory()::validate);
+                    Preconditions.checkState(operationMode.equals(Mode.ZOMBIE), "Cannot join ring without being in Zombie mode.");
```

Comment:
It's not possible AFAIK for this to ever occur, but keeping it in here just in-case